sw/device/tests/dma_inline_hashing.c

```diff
@@ -0,0 +1,171 @@
+ *
+ * This peripheral is 'direct' connected to the pads.
+ */
+void setup_pads_spi_host0(void) {
+  // set weak pull-ups for all the pads
+  dif_pinmux_pad_attr_t out_attr;
+  dif_pinmux_pad_attr_t in_attr = {
+      .slew_rate = 0,
+      .drive_strength = 0,
+      .flags = kDifPinmuxPadAttrPullResistorEnable |
+               kDifPinmuxPadAttrPullResistorUp};
+  for (uint32_t i = 0; i <= ARRAYSIZE(spi_host0_direct_pads); ++i) {
+    CHECK_DIF_OK(dif_pinmux_pad_write_attrs(&pinmux, spi_host0_direct_pads[i],
+                                            kDifPinmuxPadKindDio, in_attr,
+                                            &out_attr));
+  }
+}
+
+bool test_main(void) {
+  // Initialize the pinmux.
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_DARJEELING_PINMUX_AON_BASE_ADDR), &pinmux));
+  pinmux_testutils_init(&pinmux);
+
+  // Initialise DMA.
+  CHECK_DIF_OK(
+      dif_dma_init(mmio_region_from_addr(TOP_DARJEELING_DMA_BASE_ADDR), &dma));
+
+  // Setup pinmux if required, enable weak pull-up on relevant pads
+  setup_pads_spi_host0();  // direct
+
+  // Setup spi host configuration
+  CHECK_DIF_OK(dif_spi_host_init(
+      mmio_region_from_addr(TOP_DARJEELING_SPI_HOST0_BASE_ADDR), &spi_host));
+  init_spi_host(&spi_host, (uint32_t)kClockFreqHiSpeedPeripheralHz);
+
+  // DV sync message
+  LOG_INFO("spi host configuration complete");
+
+  dif_spi_host_segment_t host_operations[1] = {
+      {.type = kDifSpiHostSegmentTypeRx,
+       .tx = {.width = kDifSpiHostWidthStandard,
+              .buf = &received_data[TX_SIZE],
+              .length = TX_SIZE}},
+  };
+
+  // Issue the SPI transaction
+  CHECK_DIF_OK(dif_spi_host_start_transaction(
+      &spi_host, /*csid=*/0, host_operations, ARRAYSIZE(host_operations)));
+
+  // Configure the DMA to read from SPI in hardware-handshake mode
+  dif_dma_transaction_t transaction_1 = {
+      .source = {.address = TOP_DARJEELING_SPI_HOST0_BASE_ADDR +
+                            SPI_HOST_RXDATA_REG_OFFSET,
+                 .asid = kDifDmaOpentitanInternalBus},
+      .destination = {.address = (uint32_t)&received_data[0],
+                      .asid = kDifDmaOpentitanInternalBus},
+      .total_size = TX_SIZE,
+      .chunk_size = CHUNK_SIZE,
+      .width = kDifDmaTransWidth4Bytes};
+
+  dif_dma_handshake_t handshake_config = {.fifo_auto_increment = false,
+                                          .memory_auto_increment = true,
+                                          .direction_from_mem_to_fifo = false};
+
+  CHECK_DIF_OK(dif_dma_memory_range_set(&dma, TOP_DARJEELING_RAM_MAIN_BASE_ADDR,
+                                        TOP_DARJEELING_RAM_MAIN_SIZE_BYTES));
+
+  CHECK_DIF_OK(dif_dma_handshake_irq_enable(&dma, 0x2));
+  CHECK_DIF_OK(dif_dma_configure(&dma, transaction_1));
+  CHECK_DIF_OK(dif_dma_handshake_enable(&dma, handshake_config));
+
+  CHECK_DIF_OK(dif_dma_start(&dma, kDifDmaSha256Opcode));
+  CHECK_DIF_OK(dif_dma_status_poll(&dma, kDifDmaStatusDone));
```

Comment:
We probably also want to test *done* signalling by interrupt instead of polling (although I don't think this has the highest priority).

Reply:
Implemented locally. The first transfer uses IRQ signalling, the second one polling.

Reply:
Now pushed and LGTM, thanks!